### PR TITLE
Use `require` wherever possible

### LINKED ISSUE #21: require polyfill
Older versions of ComputerCraft like the one I'm running (1.74) do not support `require`. That means a lot of code in prism-core and libraries that depend on it rely heavily on `os.loadAPI` and the global environment, which is not ideal, and results in a lot of hacks to work around the fact that `os.loadAPI` will happily load the same API multiple times.

### PR DESCRIPTION
**Pull request by tomodachi94**
This PR is in a highly unfinished state. I don't recommend merging it until it's ready.

Checklist:
- [x] Move files to `.lua` extension
- [x] Have APIs (technically will be modules after this) return all of their functions inside of a table
- [x] Add a shim for `os.loadAPI` that transparently `require`s modules when `os.loadAPI()` is called
- [x] Figure out how to get `require` added on every boot
- [x] Insert the necessary code to allow `require` to work: `package.path = package.path .. ";/modules/?;/modules/?.lua;/modules/?/init.lua"`
- [ ] Test everything and fix all the bugs

Closes #21 

## Modules:
- [x] autostartup
- [x] dns
- [x] events
- [x] graph
- [x] graphmanager
- [x] input
- [x] log
- [x] messagebus
- [x] net
- [x] queuemanager
- [x] serializer
- [x] util
- [x] wire

### CODE REVIEW

**Comment by tomodachi94** [comment]
cc @danports

**Comment by tomodachi94** [comment]
@danports should we rename `util` to `prismutil` or something? I feel like it is super ambiguous.

**Comment by danports** [comment]
I agree that `util` isn't a great name. Looking at the functions in that package, I'd probably favor a name related to tables (`tablehelper`?) - admittedly still pretty generic, but then most of the API names in `prism-core` are like that. I'd suggest tackling the rename in a follow up PR though so the scope of this PR doesn't grow too large.

**Comment by tomodachi94** [comment]
Also, what are your thoughts about automatic linting? I can also setup some CI to do linting in a separate PR.

**Comment by danports** [comment]
Re: linting, my feeling is that it depends on how strict it is. Certainly a CI pipeline that checks for syntax errors at a bare minimum would be better than what we have now (nothing).

**Comment by tomodachi94** [comment]
> Re: linting, my feeling is that it depends on how strict it is. Certainly a CI pipeline that checks for syntax errors at a bare minimum would be better than what we have now (nothing).

I've got a working CI setup that uses [Selene](https://github.com/Kampfkarren/selene), a configurable linter. In its default configuration, it checks for all sorts of things, and can check even more if you configure it to do so. It also has support for knowing about custom globals (ex. ComputerCraft APIs). To top it all off, it's written in Rust.

**Comment by danports** [comment]
Sounds good. I have pushed broken code to git before, though fortunately I've caught it by other means (e.g. canary deployments) before I had to fix hundreds of broken computers! 😅 Anything to prevent that would be helpful.

**Comment by tomodachi94** [comment]
We can squash the `Convert...` commits into one when we're finished, I'm doing atomic commits in case I make a mistake.

**Comment by tomodachi94** [comment]
Alright, all of the APIs have been converted to modules.

**Comment by tomodachi94** [comment]
@danports for the `require` shim, I think we'll need to introduce a script that runs everything that is present inside of the `startup.d` folder, but as such, we'd also need to modify `railmanager` in the `prism-rails` repository to use `startup.d`.

**Comment by tomodachi94** [comment]
I've just written a script that does the heavy lifting of loading `startup.d`. It's available [here](https://gist.github.com/tomodachi94/ef2d00fa090dc593f7cdaa049db7dbe1). I've added it to this PR so we can automatically add `cc-require` on old versions. It will also help with the `os.loadAPI` shim.

(We'd need to migrate a single script in prism-rails to use this, but otherwise it wouldn't break anything.)

**Comment by tomodachi94** [comment]
Only thing that appears to be left is the `os.loadAPI` shim. (Maybe we should do some testing too.)

**Comment by danports** [comment]
I am wondering whether it would be better to handle the `require` polyfill at the package/API level, or at the ROM/BIOS level. I see you did something similar [for pastebin](https://github.com/tomodachi94/computercraft-pastebin-patch). Could we use a similar approach here? The other option would be to do something like `os.loadAPI("require")` or `dofile("cc-require")` at the top of every startup file, which seems feasible but a little inelegant. What do you think?

Also, yes, I'll do some testing on my server on this branch once it's ready.

**Comment by danports** [comment]
OK, I could see how you could do this without adding a line to the top of the files by moving all of the existing `startup` files into `startup.d` for all packages with such an entry point, which is probably better regardless, but I'm still wondering whether this should be in the library or the ROM. Given that `require` is present in the ROM in later CC versions/forks, I tend to think the polyfill should be at the ROM level too.

**Comment by tomodachi94** [comment]
> ...I'm still wondering whether this should be in the library or the ROM. Given that `require` is present in the ROM in later CC versions/forks, I tend to think the polyfill should be at the ROM level too.

I agree that it should, and cc-require is intended to be used as a resource pack that modifies ROM (but it works in non-ROM too).

My modification to cc-require detects that if `require` as a function already exists, it will abort. In other words, if we have the patch on the builtin side, ROM side and on the non-ROM side, the builtin side will take precedence, followed by ROM, followed by local.

To summarize, I think it should remain in the library but a ROM patch should be available.

**Comment by danports** [comment]
Yeah, having both library and ROM options makes a lot of sense. So the idea is that `requiem` is an optional package that can be installed by the end user on computers that would need it? We wouldn't need to take `requiem` as a dependency for every package that uses `require`, right?

**Comment by tomodachi94** [comment]
> Yeah, having both library and ROM options makes a lot of sense. So the idea is that `requiem` is an optional package that can be installed by the end user on computers that would need it? We wouldn't need to take `requiem` as a dependency for every package that uses `require`, right?

Yep, pretty much. Originally, I was going to add the `os.loadAPI` shim to `requiem`, but having it in a separate package makes a lot more sense now.

**Comment by tomodachi94** [comment]
@danports I've finished the `os.loadAPI` shim, this PR is ready for testing. I'm going to make an equivalent PR on [prism-rails](https://github.com/danports/prism-rails) that should get merged _after_ this one.

File: src/loadshim/startup.d/01-loadapi.lua [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package.path = package.path .. ";/modules/?;/modules/?.lua;/modules/?/init.lua"
+
+local old_loadapi = os.loadAPI()
```
Comment (danports) [inline src/loadshim/startup.d/01-loadapi.lua]: I think you meant to omit the `()` after `os.loadAPI` here.

Comment (tomodachi94) [reply src/loadshim/startup.d/01-loadapi.lua]: Whoops, you're right.

**Comment by danports** [comment]
Thanks! I'll take a look this weekend and see if I can test both this and the `prism-rails` PR. Some code in `amber` will likely need updating as well.

**Comment by danports** [comment]
Looking at this now...I think I need to figure out a solution for the `amber` client before I can go much further since I'll be able to pull the updated `prism-core` packages with the old `amber` but then if the `amber` client isn't compatible with those new packages, I'll have a broken installation. The `autostartup` solution is fine for daemon-type apps, but I'm not sure that's the right approach conceptually for command line apps like `amber` (well, `amber` is also a library, but let's leave that aside for now), since the `startup.d` directory mixes blocking (e.g. message loops) and non-blocking code. It seems like the options are:

1. Add `require`-specific code to `amber` (and any other command line apps that don't use `autostartup`).
2. Separate blocking and non-blocking code in `autostartup`.

What are your thoughts? I'm leaning towards 1️⃣ because I'm not sure it really makes sense to take a dependency on `autostartup` in command line apps when all we really want is `require` support (i.e. just one of the potential startup scripts). Maybe what we really want is something like a `polyfills.d` that gets loaded for command line apps and before `startup.d` for daemon apps. 🤔

File: src/autostartup/modules/autostartup.lua [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package.path = package.path .. ";/modules/?;/modules/?.lua;/modules/?/init.lua"
+local dns = require("dns")
+
```
Comment (danports) [inline src/autostartup/modules/autostartup.lua:3]: It seems like the rest of this file's contents are missing? Ditto for every other package converted to a module - it seems like only the new modules have any actual code.

File: src/startupmgr/.manifest [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+	description = "Execute all scripts in the /startup.d folder"
```
Comment (danports) [inline src/startupmgr/.manifest:2]: I'm wondering whether we should merge `startupmgr` with `autostartup`. Thoughts?

File: src/autostartup/modules/autostartup.lua (hunk 2 above)
Comment (tomodachi94) [reply src/autostartup/modules/autostartup.lua:3]: -.- Not sure what happened here. I'll try to fix this tomorrow.

File: src/startupmgr/.manifest (hunk 3 above)
Comment (tomodachi94) [reply src/startupmgr/.manifest:2]: Maybe. What are their differences?

File: src/autostartup/modules/autostartup.lua (hunk 2 above)
Comment (tomodachi94) [reply src/autostartup/modules/autostartup.lua:3]: I'll work on fix later today.

**Comment by tomodachi94** [comment]
In regards to Amber, I think we should do #1.
